### PR TITLE
feat: parse stream and generate stream

### PR DESCRIPTION
**Pull request by davidpham87**
Following https://github.com/clj-commons/clj-yaml/issues/3,
  we followed the inline implementation of
  https://github.com/metosin/muuntaja/pull/94/files
  to add the suggested feature. #

### CODE REVIEW

**Comment by borkdude** [comment]
@marcomorain @slipset Any thoughts?

**Comment by borkdude** [comment]
One could wonder why the input is an InputStream for `parse-stream` but a writer for `generate-stream`. I'm not sure what is the right answer, but it feels asymmetrical. Maybe the name of the latter function should simply be `write-yaml`? And maybe the first function should be called `read-yaml` while accepting a reader? You could then even do `(read-yaml *in*)`

**Comment by borkdude** [comment]
Another point of attention: many read functions take an `:eof` option to indicate the end of the stream. I think that would be good to add as well, or does that not make sense for reading multiple values from a stream?

**Comment by davidpham87** [comment]
Yes, I agree, it feels asymmetrical. I agree with you it would make more sense to accept a reader. I am open to the suggestion of everyone. I honestly do not know the answer, but we could follow [transit-clj](https://github.com/cognitect/transit-clj/blob/700f205781df180c3b4b251341e7f1831f9f71cb/src/cognitect/transit.clj#L319)  which also requires are reader.

**Comment by borkdude** [comment]
It seems the `.load` method supports loading from a String, InputStream or Reader:

https://github.com/asomov/snakeyaml/blob/master/src/main/java/org/yaml/snakeyaml/Yaml.java#L415-L441

There is also a `.loadAll` method which reads all objects from any of these types.

So another option could be to have a `load` and `load-all` function that works on all of these types (by doing an instance check + a cast to the right type in order to dispatch to the correct Java overloaded method). 

Similarly there is `.dump` and `.dumpAll`.

**Comment by slipset** [comment]
@marcomorain ping?

**Review by marcomorain — APPROVED** [review]
This change LGTM.

@borkdude had some questions about the API that I don't quite follow.

What API are you suggesting @borkdude ?

**Comment by borkdude** [comment]
Reiterating my previous remarks:

- [ ] The types for reading and writing differ. One takes in InputStream, the other takes a writer. Why not both streams, or writer+reader? It would be good to compare other Clojure libs (EDN, transit, JSON) for their API to offer consistency, but I think most of them take a reader/writer.

- [ ]  An `:eof` option is often supported by reading from readers/streams. E.g. `(edn/read {:eof ::eof} *in*)`.

File: src/clojure/clj_yaml/core.clj [hunk 1]
```diff
@@ -156,3 +157,18 @@
...
                  string) keywords))
+
+;; From https://github.com/metosin/muuntaja/pull/94/files
+(defn generate-stream
+  "Dump the content of data as yaml into writer."
+  [writer data & opts]
+  (.dump ^Yaml (apply make-yaml opts) (encode data) writer))
+
+(defn parse-stream
+  [^InputStream data & {:keys [unsafe mark keywords max-aliases-for-collections allow-recursive-keys allow-duplicate-keys] :or {keywords true}}]
```
Comment (borkdude) [inline src/clojure/clj_yaml/core.clj:168]: Why destructure all the keys here, but leave them implicit in generate-stream? Why not just (apply make-yaml opts) here as well?

Comment (borkdude) [inline src/clojure/clj_yaml/core.clj:168]: Why take in InputStream here and not a reader - the first argument in generate-stream is a writer.

**Comment by slipset** [comment]
@davidpham87 if you could address @borkdude's comments, we can move this forward.

**Comment by slipset** [comment]
@davidpham87 ping?

**Pull request closed without merging**